jumanji/environments/logic/sudoku/utils.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.logic.sudoku.constants import BOARD_WIDTH, BOX_IDX
+
+
+def apply_action(action: chex.Array, board: chex.Array) -> chex.Array:
+    return board.at[action[0], action[1]].set(action[2])
+
+
+def puzzle_completed(board: chex.Array) -> chex.Array:
+    """Checks that every row, column and 3x3 boxes includes all figures.
+
+    Args:
+        board: The sudoku board.
+
+    Returns:
+        condition: A `bool` indicator that validates if the puzzle is solved.
+    """
+
+    def _validate_row(row: chex.Array) -> chex.Array:
+        condition = jnp.sort(row) == jnp.arange(BOARD_WIDTH)
+        return condition.all()
+
+    condition_rows = jax.vmap(_validate_row)(board).all()
+    condition_columns = jax.vmap(_validate_row)(board.T).all()
+    condition_boxes = jax.vmap(_validate_row)(jnp.take(board, jnp.array(BOX_IDX))).all()
+    return condition_rows & condition_columns & condition_boxes
+
+
+def validate_board(board: chex.Array) -> chex.Array:
+    """Checks that every row, column and 3x3 boxes contain no duplicate number.
+
+    Args:
+        board: The sudoku board.
+
+    Returns:
+        condition: A `bool` indicator that validates a board or not.
+    """
+
+    def _validate_row(row: chex.Array) -> chex.Array:
+        return jax.nn.one_hot(row, BOARD_WIDTH).sum(axis=0).max() <= 1
+
+    condition_rows = jax.vmap(_validate_row)(board).all()
+    condition_columns = jax.vmap(_validate_row)(board.T).all()
+    condition_boxes = jax.vmap(_validate_row)(jnp.take(board, jnp.array(BOX_IDX))).all()
+
+    return condition_rows & condition_columns & condition_boxes
+
+
+def update_action_mask(action_mask: chex.Array, board: chex.Array) -> chex.Array:
+    """Updates the action mask according to the current board state.
+
+    Args:
+        action_mask: The action mask.
+        board: The sudoku board.
+
+    Returns:
+        The updated action mask.
+    """
+    row_mask = 1 - jax.nn.one_hot(board, BOARD_WIDTH).any(axis=1) * 1
+    column_mask = 1 - jax.nn.one_hot(board.T, BOARD_WIDTH).any(axis=1) * 1
+
+    boxes = board.reshape(BOARD_WIDTH**2).take(jnp.array(BOX_IDX))
```

Comment:
I don't like that very much in general, because it forces the Jax module to be initialized whenever jumanji is imported, I prefer storing numpy arrays then casting them to jax arrays. But if you have a strong opinion on this we might change.

Reply:
I think both codes are completely equivalent. And I think this also works the same:
```python
    boxes = board.reshape(BOARD_WIDTH**2).take(BOX_IDX)
```
So, I'm happy with either.

Reply:
Good point @Egiob ! no strong opinion you can keep it like that :) 
Also like the alternative by @clement-bonnet

Reply:
Now I define BOX_IDX with np.array, so I don't need it to cast it explicitely in Jax afterwards